# In-Class Coding Lab: Data Analysis with Pandas

In this lab, we will perform a data analysis on the **RMS Titanic** passenger list. The RMS Titanic is one of the most famous ocean liners in history. On April 15, 1912 it sank after colliding with an iceberg in the North Atlantic Ocean. To learn more, read here: https://en.wikipedia.org/wiki/RMS_Titanic 

Our goal today is to perform a data analysis on a subset of the passenger list. We're looking for insights as to which types of passengers did and didn't survive. Women? Children? 1st Class Passengers? 3rd class? Etc. 

I'm sure you've heard the expression often said during emergencies: "Women and Children first" Let's explore this data set and find out if that's true!

Before we begin you should read up on what each of the columns mean in the data dictionary. You can find this information on this page: https://www.kaggle.com/c/titanic/data 


## Loading the data set

First we load the dataset into a Pandas `DataFrame` variable. The `sample(10)` method takes a random sample of 10 passengers from the data set.

In [2]:
import pandas as pd
import numpy as np

# this turns off warning messages
import warnings
warnings.filterwarnings('ignore')

passengers = pd.read_csv('ICCL-titanic.csv')
passengers.sample(10)

PassengerId  Survived  Pclass  \
144          145         0       2   
183          184         1       2   
458          459         1       2   
696          697         0       3   
801          802         1       2   
196          197         0       3   
671          672         0       1   
155          156         0       1   
723          724         0       2   
357          358         0       2   

                                            Name     Sex   Age  SibSp  Parch  \
144                   Andrew, Mr. Edgardo Samuel    male  18.0      0      0   
183                    Becker, Master. Richard F    male   1.0      2      1   
458                          Toomey, Miss. Ellen  female  50.0      0      0   
696                             Kelly, Mr. James    male  44.0      0      0   
801  Collyer, Mrs. Harvey (Charlotte Annie Tate)  female  31.0      1      1   
196                          Mernagh, Mr. Robert    male   NaN      0      0   
671                       Davidson, Mr. Thornton    male  31.0      1      0   
155                  Williams, Mr. Charles Duane    male  51.0      0      1   
723                      Hodges, Mr. Henry Price    male  50.0      0      0   
357                    Funk, Miss. Annie Clemmer  female  38.0      0      0   

           Ticket     Fare Cabin Embarked  
144        231945  11.5000   NaN        S  
183        230136  39.0000    F4        S  
458  F.C.C. 13531  10.5000   NaN        S  
696        363592   8.0500   NaN        S  
801    C.A. 31921  26.2500   NaN        S  
196        368703   7.7500   NaN        Q  
671    F.C. 12750  52.0000   B71        S  
155      PC 17597  61.3792   NaN        C  
723        250643  13.0000   NaN        S  
357        237671  13.0000   NaN        S

## How many survived?

One of the first things we should do is figure out how many of the passengers in this data set survived. Let's start with isolating just the `'Survivied'` column into a series:

In [3]:
passengers['Survived'].sample(10)

698    0
399    1
334    1
667    0
105    0
291    1
415    0
512    1
487    0
503    0
Name: Survived, dtype: int64

There's too many to display so we just display a random sample of 10 passengers. 

- 1 means the passenger survivied
- 0 means the passenger died

What we really want is to count the number of survivors and deaths. We do this by querying the `value_counts()` of the `['Survived']` column, which returns a `Series` of counts, like this:

In [4]:
passengers['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

Only 342 passengers survived, and 549 perished. Let's observe this same data as percentages of the whole. We do this by adding the `normalize=True` named argument to the `value_counts()` method.

In [5]:
passengers['Survived'].value_counts(normalize=True)

0    0.616162
1    0.383838
Name: Survived, dtype: float64

**Just 38% of passengers in this dataset survived.**

### Now you Try it!

**FIRST** Write a Pandas expression to display counts of males and female passengers using the `Sex` variable:

In [6]:
# todo write code here
import pandas as pd
import numpy as np
import warnings 
warnings.filterwarnings('ignore')

passengers = pd.read_csv('ICCL-titanic.csv')
passengers.sample(10)
passengers['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

**NEXT** Write a Pandas expression to display male /female passenger counts as a percentage of the whole number of passengers in the data set.

In [7]:
# todo write code here
passengers['Sex'].value_counts(normalize=True)

male      0.647587
female    0.352413
Name: Sex, dtype: float64

If you got things working, you now know that **35% of passengers were female**.

## Who survivies? Men or Women?

We now know that 35% of the passengers were female, and 65% we male. 

**The next think to think about is how do survivial rates affect these numbers? **

If the ratio is about the same for surviviors only, then we can conclude that your **Sex** did not play a role in your survival on the RMS Titanic. 

Let's find out.

In [8]:
survivors = passengers[passengers['Survived'] ==1]
survivors['PassengerId'].count()

342

Still **342** like we discovered originally. Now let's check the **Sex** split among survivors only:

In [9]:
survivors['Sex'].value_counts()

female    233
male      109
Name: Sex, dtype: int64

WOW! That is a huge difference! But you probably can't see it easily. Let's represent it in a `DataFrame`, so that it's easier to visualize:

In [10]:
sex_all_series = passengers['Sex'].value_counts()
sex_survivor_series = survivors['Sex'].value_counts()

sex_comparision_df = pd.DataFrame({ 'AllPassengers' : sex_all_series, 'Survivors' : sex_survivor_series })
sex_comparision_df['SexSurvivialRate'] = sex_comparision_df['Survivors'] / sex_comparision_df['AllPassengers']
sex_comparision_df

AllPassengers  Survivors  SexSurvivialRate
female            314        233          0.742038
male              577        109          0.188908

 **So, females had a 74% survival rate. Much better than the overall rate of 38%**
 
We should probably briefly explain the code above. 

- The first two lines get a series count of all passengers by Sex (male / female) and count of survivors by sex
- The third line creates DataFrame. Recall a pandas dataframe is just a dict of series. We have two keys 'AllPassengers' and 'Survivors'
- The  fourth line creates a new column in the dataframe which is just the survivors / all passengers to get the rate of survival for that Sex.

## Feature Engineering: Adults and Children

Sometimes the variable we want to analyze is not readily available, but can be created from existing data. This is commonly referred to as **feature engineering**. The name comes from machine learning where we use data called *features* to predict an outcome. 

Let's create a new feature called `'AgeCat'` as follows:

- When **Age** <=18 then 'Child'
- When **Age** >18 then 'Adult'

This is easy to do in pandas. First we create the column and set all values to `np.nan` which means 'Not a number'. This is Pandas way of saying no value. Then we set the values based on the rules we set for the feature.

In [11]:
passengers['AgeCat'] = np.nan # Not a number
passengers['AgeCat'][ passengers['Age'] <=18 ] = 'Child'
passengers['AgeCat'][ passengers['Age'] > 18 ] = 'Adult'
passengers.sample(5)

PassengerId  Survived  Pclass  \
241          242         1       3   
710          711         1       1   
124          125         0       1   
535          536         1       2   
391          392         1       3   

                                                 Name     Sex   Age  SibSp  \
241                    Murphy, Miss. Katherine "Kate"  female   NaN      1   
710  Mayne, Mlle. Berthe Antonine ("Mrs de Villiers")  female  24.0      0   
124                       White, Mr. Percival Wayland    male  54.0      0   
535                            Hart, Miss. Eva Miriam  female   7.0      0   
391                            Jansson, Mr. Carl Olof    male  21.0      0   

     Parch        Ticket     Fare Cabin Embarked AgeCat  
241      0        367230  15.5000   NaN        Q    NaN  
710      0      PC 17482  49.5042   C90        C  Adult  
124      1         35281  77.2875   D26        S  Adult  
535      2  F.C.C. 13529  26.2500   NaN        S  Child  
391      0        350034   7.7958   NaN        S  Adult

Let's get the count and distrubutions of Adults and Children on the passenger list.

In [12]:
passengers['AgeCat'].value_counts()

Adult    575
Child    139
Name: AgeCat, dtype: int64

And here's the percentage as a whole:

In [13]:
passengers['AgeCat'].value_counts(normalize=True)

Adult    0.805322
Child    0.194678
Name: AgeCat, dtype: float64

So close to **80%** of the passengers were adults. Once again let's look at the ratio of `AgeCat` for survivors only. If your age has no bearing of survivial, then the rates should be the same. 

Here's the counts of Adult / Children among the survivors only:

In [14]:
survivors = passengers[passengers['Survived'] ==1]
survivors['AgeCat'].value_counts()

Adult    220
Child     70
Name: AgeCat, dtype: int64

### Now You Try it!

Calculate the `AgeCat` survival rate, similar to how we did for the `SexSurvivalRate`. 

In [19]:
agecat_all_series = passengers['AgeCat'].value_counts()
agecat_survivor_series = survivors['AgeCat'].value_counts()
agecat_comparison_df=pd.DataFrame({'All passengers':agecat_all_series,'Survivors':agecat_survivor_series})
agecat_comparison_df['sexSurvivalRate']=agecat_comparison_df['Survivors']/agecat_comparison_df['All passengers']
agecat_comparison_df
# todo make a data frame, add AgeCatSurvivialRate column, display dataframe 


All passengers  Survivors  sexSurvivalRate
Adult             575        220         0.382609
Child             139         70         0.503597

**So, children had a 50% survival rate, better than the overall rate of 38%**

## So, women and children first?

It looks like the RMS really did have the motto: "Women and Children First."

Here's our insights. We know:

- If you were a passenger, you had a 38% chance of survival.
- If you were a female passenger, you had a 74% chance of survival.
- If you were a child passenger, you had a 50% chance of survival. 


### Now you try it for Passenger Class

Repeat this process for `Pclass` The passenger class variable. Display the survival rates for each passenger class. What does the information tell you about passenger class and survival rates?

I'll give you a hint... "Money Talks"


In [25]:
# todo: repeat the analysis in the previous cell for Pclass 
pc_all_series=passengers['Pclass'].value_counts()
pc_survivors=survivors['Pclass'].value_counts()

pc_comparison_df=pd.DataFrame({'All passengers':pc_all_series,'Survivors':pc_survivors})
pc_comparison_df['Pclasssurvivalrate']=pc_comparison_df['Survivors']/pc_comparison_df['All passengers']
pc_comparison_df


All passengers  Survivors  Pclasssurvivalrate
1             216        136            0.629630
2             184         87            0.472826
3             491        119            0.242363